src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -1665,20 +1669,21 @@ static inline freeze_result freeze_epilog(ContinuationWrapper& cont) {
   return freeze_ok;
 }
 
-static freeze_result freeze_epilog(JavaThread* thread, ContinuationWrapper& cont, freeze_result res) {
+static freeze_result freeze_epilog(JavaThread* current, ContinuationWrapper& cont, freeze_result res) {
   if (UNLIKELY(res != freeze_ok)) {
-    JFR_ONLY(thread->set_last_freeze_fail_result(res);)
+    JFR_ONLY(current->set_last_freeze_fail_result(res);)
     verify_continuation(cont.continuation());
     log_develop_trace(continuations)("=== end of freeze (fail %d)", res);
     return res;
   }
 
-  JVMTI_ONLY(jvmti_yield_cleanup(thread, cont)); // can safepoint
+  JVMTI_ONLY(jvmti_yield_cleanup(current, cont)); // can safepoint
   return freeze_epilog(cont);
 }
 
-static freeze_result preempt_epilog(ContinuationWrapper& cont, freeze_result res, frame& old_last_frame) {
+static freeze_result preempt_epilog(JavaThread* current, ContinuationWrapper& cont, freeze_result res, frame& old_last_frame) {
   if (UNLIKELY(res != freeze_ok)) {
+    JFR_ONLY(current->set_last_freeze_fail_result(res);)
```

Comment:
Note that we don’t need to save the result when unmounting from the VM, i.e the preempt case. We return it from `Continuation::try_preempt()` so it is already available when we call `post_vthread_pinned_event()` [1][2]. When unmounting from Java we do need to save it because the call to `post_vthread_pinned_event()` will be done later in `VirtualThread.postPinnedEvent()`, where the return value from the freeze call is not available anymore. 

[1] https://github.com/openjdk/jdk/blob/30b8bbe255c1653d25961c3fe79096b2b8d62bd6/src/hotspot/share/runtime/objectMonitor.cpp#L603
[2] https://github.com/openjdk/jdk/blob/30b8bbe255c1653d25961c3fe79096b2b8d62bd6/src/hotspot/share/runtime/objectMonitor.cpp#L1824

Reply:
I see. Thanks for making us aware.

Reply:
The JFR event support changed a bit near the end so looks like this slipped through. For the non-preempt cases, the _last_freeze_fail_XXX values must be set so that the later call to record an event has the last reason/time. So I think the change look good. I see the failure was in legacy mode so more likely to run into this.